### ISSUE TITLE
A syntax for string diagrams

### ISSUE THREAD

**Issue by mroman42** [ohnosequences/monochord#2]
**What syntax should we use to represent string diagrams internally?**

My first idea is to write the diagrams in terms of vertical and horizontal compositions of natural transformations. Every natural transformation can be seen as a rectangle of the diagram with a number of
"inputs" and "outputs". And those rectangles can be composed vertically and horizontally.

As an **example**, let's take the string diagram for an adjuntion
![adjuntion](https://cloud.githubusercontent.com/assets/5337877/8849962/5043bd7a-3144-11e5-8724-7c0049d0d7c9.png)

This can be seen as the vertical composition of:
- The horizontal composition of these two:  
  ![f](https://cloud.githubusercontent.com/assets/5337877/8850060/33dff6b6-3145-11e5-8389-868d93a43873.png)
  ![e](https://cloud.githubusercontent.com/assets/5337877/8850062/35da0542-3145-11e5-926e-8a51af506d4b.png)
- And the horizontal composition of these two:  
  ![n](https://cloud.githubusercontent.com/assets/5337877/8850076/476a82e6-3145-11e5-8964-f9c6177b5fe6.png)
  ![f](https://cloud.githubusercontent.com/assets/5337877/8850060/33dff6b6-3145-11e5-8389-868d93a43873.png)

Where the identity transformations have 1 input and 1 output, while _epsilon_ and _eta_ have 2 inputs and 1 output and 1 input and 2 outputs respectively. We are not drawing the identities. In the proposed syntax this would translate to:

> (F \* ε) o (η \* F)

which is equal to `F` (the identity natural transformation in the functor **F**) if we read it as an equation.

**Comment by eparejatobes** [comment]
Yup. The fun starts when you need to work with all that is still implicit in what you write there: ε has arity (2,1), these input/output things are typed, ...

Recommendations:
- [Open graphs and monoidal theories](http://arxiv.org/pdf/1011.4114v1.pdf) which is actually the same technology (cospans of labeled (monoidal) graphs with discrete source and target) as in
- [Calculating colimits compositionally](https://irinsubria.uninsubria.it/retrieve/handle/11383/1746288/2116/ugo-volume.pdf)

The actual reason all this works is that cospans (or spans) contain a lot of free monoidal blablabla constructions.

**Comment by eparejatobes** [comment]
This is a good read for getting some context
- [Graphical languages - RFC Walters](http://rfcwalters.blogspot.com.es/2009/08/graphical-languages-27-april-2009.html)

**Comment by mroman42** [comment]
I see. I'm going to read the papers.
